### DISCUSSION TITLE
zone trigger (or proximity)

### DISCUSSION THREAD

**Discussion by stefanuytterhoeven** [custom-components/pyscript#276] [Q&A] [answered]
in HA there's a platform : zone (zone trigger)
how can i do this in pyscript? 
I would like to know when a person is leaving/entering a zone, (zone.home, or another defined zone)

And how can i work with "proximity" integration?

Somebody has examples?

regards,
stefan

**Comment by dlashua** [comment]
entities in the `zone` and `proximity` domain work just like any other.

Here's some sample code:
```
@state_trigger('True or proximity.daniel')
def whatever():
  if proximity.daniel > 50:
    log.info('Daniel is more than 50 away')
  else:
    log.info(Daniel is less than 50 away')
```

**Reply by stefanuytterhoeven** [reply]
Aha ! Thanks, Will try that. Not easy to test :)
Thx!

**Reply by stefanuytterhoeven** [reply]
What about zone trigger?
likes this automation:
automation:
  trigger:
    - platform: zone
      entity_id: person.paulus
      zone: zone.home
      # Event is either enter or leave
      event: enter # or "leave"

How can I do this pyscript? 
I would like to know when someone enters/leaves (the event:enter or event:leave) a zone. 
A "state trigger" on "zone" will not work, since "zone" always contains the value "zoning". But, according to the docs, there is a "zone trigger", but I don't know how to use a "zone trigger" in pyscript.

**Reply by dlashua** [reply]
I'm not sure. I use the `person` entity to determine the zone a person is in.

If you need to use `zone` for whatever reason, as you indicated, pyscript doesn't support it out of the box in the same way that Home Assistant Automations do. You can read the source code for Zone and determine which events it might trigger (since the Docs don't indicate anything in this regard). If it does trigger events, you can use `@event_trigger` to see those. If it doesn't do that either, you can use the `hass` object in pyscript to implement the Home Assistant Zone Automation Trigger using, essentially, the same code contained in the `zone` component code.

**Reply by dlashua** [reply]
Based on the [code for the zone trigger](https://github.com/home-assistant/core/blob/939fde0a500f599fac9f9c5cd2befa8b75f45475/homeassistant/components/zone/trigger.py#L36), it's just listening for changes in the `entity_id` of the trigger -- the `person.` entity. If `event` is "enter" it triggers when the state changes TO the zone. When it is "leave" it triggers when the state changes FROM that zone.

**Reply by stefanuytterhoeven** [reply]
ha, ok... that explains why the automation also uses the "person" entity.
I already have a app that sends a notification (for testing) when leaving and entering a zone 

part of the app;

@state_trigger(f"{config['person_entity']}")     
    def send_notification1(trigger_type=None, var_name=None, value=None, old_value=None):
      => send notifcation with old_value, value. 

and I got messages, like old_value="home" , value "not_home"
then value "friends_home", old_value="not_home"

but this is all done with a state trigger... 
So, you say, changes in the entity_id of the trigger -- the person....   so, it will also be a "state trigger" I guess.
Don't know yet where to get the "event". 

Thanks anyway! I already searched around, and eventually asked it here. Maybe there was a simple solution... I'll have to look further.
But, now I'll look further. 
:) thx!

**Comment by dlashua** [comment, marked as answer]
That's exactly how `person` entities work. They indicate the name of the zone the person is in, or, if they aren't in any particular zone, then they are set to `not_home`.

There is no event (as best as I can tell from reading the code). Just the state of the `person.whatever` changing to/from the name of the zone.
